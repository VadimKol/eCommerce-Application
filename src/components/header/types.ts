export interface HeaderProps {
  name?: string;
}
