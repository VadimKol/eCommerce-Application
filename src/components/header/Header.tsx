import { useContext } from 'react';
import { Link, useLocation, useNavigate } from 'react-router-dom';
import { LoginContext } from '@/contexts/login-context.tsx';
import styles from './styles.module.scss';

enum NavigationPaths {
  HOME = '/',
  CATALOG = '/catalog',
  ABOUT = '/about',
}

enum ActionPaths {
  LOGIN = '/login',
  REGISTER = '/register',
}

export function Header(): JSX.Element {
  const { isAuthenticated, setIsAuthenticated } = useContext(LoginContext);
  const location = useLocation();
  const navigate = useNavigate();

  const handleLogout = (): void => {
    setIsAuthenticated(false);

    if (location.pathname !== String(NavigationPaths.HOME)) {
      navigate(NavigationPaths.HOME);
    }
  };

  const isCurrentLink = (path: NavigationPaths): boolean => location.pathname === String(path);
  const shouldShowRegisterLink = location.pathname !== String(ActionPaths.REGISTER);
  const shouldShowLoginLink = location.pathname !== String(ActionPaths.LOGIN);

  return (
    <header className={styles.header}>
      <div className={styles.logo}>
        <Link to={NavigationPaths.HOME}>
          <img src="/logo.png" alt="Geek Store logo" />
        </Link>
      </div>
      <nav className={styles.navigation}>
        <ul className={styles.navList}>
          {Object.keys(NavigationPaths).map((path) => (
            <li key={path}>
              {isCurrentLink(NavigationPaths[path as keyof typeof NavigationPaths]) ? (
                <span className={`${styles.navLink} ${styles.navLinkCurrent}`}>{path}</span>
              ) : (
                <Link className={styles.navLink} to={NavigationPaths[path as keyof typeof NavigationPaths]}>
                  {path}
                </Link>
              )}
            </li>
          ))}
        </ul>
      </nav>
      <div className={styles.actions}>
        {isAuthenticated ? (
          <button className={styles.logoutButton} type="button" onClick={handleLogout}>
            Logout
          </button>
        ) : (
          <>
            {shouldShowLoginLink && (
              <Link to={ActionPaths.LOGIN} className={styles.loginLink}>
                Login
              </Link>
            )}
            {shouldShowRegisterLink && (
              <Link to={ActionPaths.REGISTER} className={styles.registerLink}>
                Register
              </Link>
            )}
          </>
        )}
      </div>
    </header>
  );
}
