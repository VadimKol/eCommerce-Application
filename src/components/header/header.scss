* {
  box-sizing: content-box;
}

header {
  display: flex;
  flex-flow: row nowrap;
  justify-content: space-between;
  align-items: baseline;
  width: 100%;
  height: 13%;
  padding: 0 10%;
  font-size: 19px;
  color: var(--constract-text-color);
  background: var(--main-color-first);
}

h1 {
  margin: 25px 0;
  font-size: 50px;
  text-align: center;
  text-shadow:
    -1px -1px 0 var(--main-color-second),
    1px -1px 0 var(--main-color-second),
    -1px 1px 0 var(--main-color-second),
    1px 1px 0 var(--main-color-second);
  color: var(--main-color-first);
}
