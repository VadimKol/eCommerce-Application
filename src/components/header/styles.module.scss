.header {
  position: relative;
  width: 100%;
  background-image: linear-gradient(to bottom, var(--primary-color), #073283);
}

.container {
  display: grid;
  grid-template-columns: 1fr max-content 1fr;
  align-items: center;
  width: 100%;
  max-width: var(--max-width);
  margin-inline: auto;
  padding: 9px 10px 10px;
  font-size: 20px;
  color: var(--text-color);

  @media (width <= 768px) {
    grid-template-columns: 1fr;
  }
}

.wordmark {
  display: block;
  width: fit-content;
  font-family: var(--secondary-font);
  font-size: 34px;
  font-weight: 700;
  text-decoration: none;
  text-transform: uppercase;
  transition: -webkit-text-stroke-color 0.6s ease;
  -webkit-text-fill-color: var(--primary-color);
  -webkit-text-stroke-width: 1px;
  -webkit-text-stroke-color: var(--secondary-color);

  &:hover,
  &:focus-visible {
    @media (hover: hover) {
      -webkit-text-stroke-color: var(--tertiary-color);
    }
  }
}
