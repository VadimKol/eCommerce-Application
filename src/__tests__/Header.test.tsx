import { render } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';

import { Header } from '@/components/header/Header';

jest.mock('./styles.module.scss', () => ({
  header: 'header',
  container: 'container',
  wordmark: 'wordmark',
}));

describe('Header Component', () => {
  test('renders correctly and matches snapshot', () => {
    const { asFragment } = render(
      <BrowserRouter>
        <Header />
      </BrowserRouter>,
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
