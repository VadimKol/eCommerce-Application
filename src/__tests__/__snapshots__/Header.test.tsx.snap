// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Component renders correctly and matches snapshot 1`] = `
<DocumentFragment>
  <header
    class="header"
  >
    <div>
      <div
        class="bm-overlay"
        style="position: fixed; z-index: 1000; width: 100%; height: 100%; background: rgba(0, 0, 0, 0.3); opacity: 0; transform: translate3d(100%, 0, 0); transition: opacity 0.3s, transform 0s 0.3s;"
      />
      <div>
        <div
          class="bm-burger-button"
          style="z-index: 1000;"
        >
          <button
            id="react-burger-menu-btn"
            style="position: absolute; left: 0px; top: 0px; z-index: 1; width: 100%; height: 100%; margin: 0px; padding: 0px; font-size: 0px; background: transparent; cursor: pointer;"
            type="button"
          >
            Open Menu
          </button>
          <span>
            <span
              class="bm-burger-bars"
              style="position: absolute; height: 20%; left: 0px; right: 0px; top: 0%; opacity: 1;"
            />
            <span
              class="bm-burger-bars"
              style="position: absolute; height: 20%; left: 0px; right: 0px; top: 40%; opacity: 1;"
            />
            <span
              class="bm-burger-bars"
              style="position: absolute; height: 20%; left: 0px; right: 0px; top: 80%; opacity: 1;"
            />
          </span>
        </div>
      </div>
      <div
        aria-hidden="true"
        class="bm-menu-wrap"
        id=""
        style="position: fixed; right: 0px; z-index: 1100; width: 280px; height: 100%; transform: translate3d(100%, 0, 0); transition: all 0.5s;"
      >
        <div
          class="bm-menu"
          style="height: 100%; box-sizing: border-box; overflow: auto;"
        >
          <div
            class="bm-item-list"
            style="height: 100%;"
          >
            <nav
              class=""
            >
              <ul>
                <li>
                  <span
                    class=""
                  >
                    HOME
                  </span>
                </li>
                <li>
                  <a
                    href="/catalog"
                  >
                    CATALOG
                  </a>
                </li>
                <li>
                  <a
                    href="/about"
                  >
                    ABOUT
                  </a>
                </li>
              </ul>
            </nav>
            <ul
              class=""
            >
              <li>
                <a
                  class=""
                  href="/login"
                >
                  Login
                </a>
              </li>
              <li>
                <a
                  class=""
                  href="/register"
                >
                  Register
                </a>
              </li>
            </ul>
          </div>
        </div>
        <div>
          <div
            class="bm-cross-button"
            style="position: absolute; width: 24px; height: 24px; right: 8px; top: 8px;"
          >
            <button
              id="react-burger-cross-btn"
              style="position: absolute; left: 0px; top: 0px; z-index: 1; width: 100%; height: 100%; margin: 0px; padding: 0px; font-size: 0px; background: transparent; cursor: pointer;"
              tabindex="-1"
              type="button"
            >
              Close Menu
            </button>
            <span
              style="position: absolute; top: 6px; right: 14px;"
            >
              <span
                class="bm-cross"
                style="position: absolute; width: 3px; height: 14px; transform: rotate(45deg);"
              />
              <span
                class="bm-cross"
                style="position: absolute; width: 3px; height: 14px; transform: rotate(-45deg);"
              />
            </span>
          </div>
        </div>
      </div>
    </div>
    <div
      class="container"
    >
      <a
        class="wordmark"
        href="/"
      >
        Geek store
      </a>
      <nav
        class=""
      >
        <ul>
          <li>
            <span
              class=""
            >
              HOME
            </span>
          </li>
          <li>
            <a
              href="/catalog"
            >
              CATALOG
            </a>
          </li>
          <li>
            <a
              href="/about"
            >
              ABOUT
            </a>
          </li>
        </ul>
      </nav>
      <ul
        class=""
      >
        <li>
          <a
            class=""
            href="/login"
          >
            Login
          </a>
        </li>
        <li>
          <a
            class=""
            href="/register"
          >
            Register
          </a>
        </li>
      </ul>
    </div>
  </header>
</DocumentFragment>
`;
